### PR TITLE
Dynamic and custom configuration of client configs

### PR DESCRIPTION
**Pull request by phutchins**
Allows users to more granularly and dynamically control the client config created.

### CODE REVIEW

**Comment by flaccid** [comment]
Can we have the tests pass first please?

**Comment by flaccid** [comment]
Any reason this wouldn't simply use the existing LWRP and `config` attribute https://github.com/xhost-cookbooks/openvpn/blob/master/providers/conf.rb ?

**Comment by flaccid** [comment]
@phutchins just circling back on this one - keen to hear your thoughts, perhaps I am misunderstanding.. let me know please.

**Comment by phutchins** [comment]
Doing it this way allows us to set different parameters on the client and server side. The changes also allow for flags vs key/value config options.

**Comment by flaccid** [comment]
I'd like to avoid going back to separate keys for client and server configuration. I think the requirement for the client template to be not hard coded is what we should be looking at specifically. I think there must be a more elegant way to do this. I will have a play soon.

**Comment by phutchins** [comment]
What is the reason for not having separate keys for client and server? That seems like an arbitrary restriction. 

I definitely considered ways to merge flags and config values but the options seemed more messy than this way...

**Comment by devsibwarra** [comment]
It would be nice to see a generic template file, without any static lines, that takes the config hash and converts that into a file that OpenVPN can consume.  It looks like your clients.conf template change has part of this already, but I'd like seeing the other hard-coded lines moved out

Instead of adding a `client_flags` setting, you can just add the desired options to your `config` hash like so
```
  'config' => {
    'comp-lzo' => '',
    'persist-key' => '',
    'persist-tun' => '',
    'nobind' => '',
    ...
  }
```

I do this with our server config to add the `opt-verify` flag, but this should work just fine with your tweaks to the client template.

**Comment by phutchins** [comment]
Definitely. Thats pretty much what I was going for. When all of the values are in the same config object however it will drop them into the template out of order and I desired a config with the single word entries (flags) together, and the config params together. That is more of a organization and readability thing over functionality so I could see it being useful both ways...

The functionality in the client.conf template does already exist to parse an empty value as a single word entry so this wouldn't be an issue.

We could merge client_flags with client_config, however, I want the ability to have different config values, naming, etc... between my server config and client config so I want those separate. They could default to the same but I dynamically name these certs for the client and the client config needs to be able to reflect that.

Some of the values, proto, for instance, will always be shared between server and client so I can individually add those to the client_config object setting them to the servers value. This way it can be changed by changing only the value in 'config'.

Thoughts?

**Comment by phutchins** [comment]
@devsibwarra I just pushed some changes in line with our discussion. Take a look and let me know your thoughts... Tests aren't passing on my machine so we'll figure that out once we land on a direction.

File: attributes/default.rb [hunk 1]
```diff
@@ -75,8 +77,9 @@
 default['openvpn']['config']['persist-key']     = ''
 default['openvpn']['config']['persist-tun']     = ''
 default['openvpn']['config']['comp-lzo']        = ''
+default['openvpn']['config']['nobind']          = ''
 
-default['openvpn']['config']['ca']              = node['openvpn']['signing_ca_cert']
+default['openvpn']['config']['ca']              = "#{node['openvpn']['signing_ca_cert']}/#{node['openvpn']['ca_cert_filename']}"
```
Comment (devsibwarra) [inline attributes/default.rb]: Looks like this will be set by default to `/etc/openvpn/keys/ca.crt/ca.crt`.  You can probably revert this change and leave it assigned to `node[openvpn][signing_ca_cert]`

Comment (devsibwarra) [inline attributes/default.rb]: Looks like this will be set by default to `/etc/openvpn/keys/ca.crt/ca.crt`.  You can probably revert this change and leave it assigned to `node[openvpn][signing_ca_cert]`

File: recipes/users.rb [hunk 2]
```diff
@@ -48,10 +48,15 @@ def chef_solo_search_installed?
       not_if { ::File.exist?("#{node['openvpn']['key_dir']}/#{u['id']}.crt") }
     end
 
+    node.set['openvpn']['client_config']['cert'] = "#{u['id']}.crt"
+    node.set['openvpn']['client_config']['key']  = "#{u['id']}.key"
```
Comment (devsibwarra) [inline recipes/users.rb]: From http://www.rubydoc.info/gems/chef/Chef/Node#set-instance_method
```
def set
  Chef.log_deprecation("node.set is deprecated and will be removed in Chef 14, please use node.default/node.override (or node.normal only if you really need persistence)")
  normal
end
```

File: attributes/default.rb [hunk 3]
```diff
@@ -32,28 +32,30 @@
...
 
 # Cookbook attributes
-default['openvpn']['key_dir']         = '/etc/openvpn/keys'
-default['openvpn']['signing_ca_key']  = "#{node['openvpn']['key_dir']}/ca.key"
-default['openvpn']['signing_ca_cert'] = "#{node['openvpn']['key_dir']}/ca.crt"
-default['openvpn']['user_query']      = '*:*'
+default['openvpn']['key_dir']          = '/etc/openvpn/keys'
+default['openvpn']['ca_key_filename']  = 'ca.key'
+default['openvpn']['signing_ca_key']   = "#{node['openvpn']['key_dir']}/#{node['openvpn']['ca_key_filename']}"
+default['openvpn']['ca_cert_filename'] = "ca.crt"
```
Comment (devsibwarra) [inline attributes/default.rb]: rubocop will be happy if you use single-quotes for a string that has nothing to parse/replace

**Comment by devsibwarra** [comment]
Since there are numerous defaults being set for the client config, there should be a way for a user to bypass/ignore some of the unwanted default settings, rather than have a wrapper script do `node.rm_default(...)`.

I've done this in the past with other cookbooks by setting the value to `nil` in my environment/role configs and skipping that value in the template loop.

File: recipes/users.rb (hunk 2 above)
Comment (devsibwarra) [inline recipes/users.rb]: Expanded the code here to see where `u['id']` was coming from and I'm thinking this should be changed as you're doing a global `node.set` inside a dynamic list of users.  IIRC, this will cause `node[openvpn][client_config][cert]` to be set to the last user processed and all templates will end up using that value.  Probably want to set the values in a local hash, then merge that into `node[openvpn][client_config]` when the template is evaluated, to avoid impacting global settings.

Example:
```ruby
local_config = {
  'cert' => "#{u['id']}.crt",
  'key' => "#{u['id']}.key"
}
local_config.merge!(node['openvpn']['client_config'])

template ... do
  variables(
    config: local_config
  )
end
```

If someone sets their own custom cert/key, according to [Ruby Hash .merge](https://ruby-doc.org/core-2.3.1/Hash.html#method-i-merge-21), it will override the local_config value, which seems correct

**Comment by damacus** [comment]
I'm going to close this one down now. If you want to re-open this please feel free! 

Thanks! 
Dan
